libmachine/mcnutils/crc.go

```diff
@@ -0,0 +1,227 @@
+		Proxy:             http.ProxyFromEnvironment,
+		Dial:              defaultTimeout,
+	}
+
+	return &http.Client{
+		Transport: &transport,
+	}
+}
+
+// releaseGetter is a client that gets release information of a product and downloads it.
+type releaseGetter interface {
+	// filename returns filename of the product.
+	filename() string
+
+	// download downloads a file from the given dlURL and saves it under dir.
+	download(dir, file, dlURL string) error
+}
+
+// disk is an disk path.
+type disk interface {
+	// path returns the path of the Disk.
+	path() string
+	// exists reports whether the Disk exists.
+	exists() bool
+}
+
+// crcDisk represents a CRC disk Image. It implements the Disk interface.
+type crcDisk struct {
+	// path of Disk ISO
+	commonDiskPath string
+}
+
+func (b *crcDisk) path() string {
+	if b == nil {
+		return ""
+	}
+	return b.commonDiskPath
+}
+
+func (b *crcDisk) exists() bool {
+	if b == nil {
+		return false
+	}
+
+	_, err := os.Stat(b.commonDiskPath)
+	return !os.IsNotExist(err)
+}
+
+func removeFileIfExists(name string) error {
+	if _, err := os.Stat(name); err == nil {
+		if err := os.Remove(name); err != nil {
+			return fmt.Errorf("Error removing temporary download file: %s", err)
+		}
+	}
+	return nil
+}
+
+// crcReleaseGetter implements the releaseGetter interface for getting the Disk image.
+type crcReleaseGetter struct {
+	diskFilename string
+}
+
+func (b *crcReleaseGetter) filename() string {
+	if b == nil {
+		return ""
+	}
+	return b.diskFilename
+}
+
+func (*crcReleaseGetter) download(dir, file, isoURL string) error {
+	u, err := url.Parse(isoURL)
+
+	var src io.ReadCloser
+	if u.Scheme == "file" || u.Scheme == "" {
+		s, err := os.Open(u.Path)
+		if err != nil {
+			return err
+		}
+
+		src = s
+	} else {
+		client := getClient()
+		s, err := client.Get(isoURL)
+		if err != nil {
+			return err
+		}
+
+		src = &ReaderWithProgress{
+			ReadCloser:     s.Body,
+			out:            os.Stdout,
+			expectedLength: s.ContentLength,
+		}
+	}
+
+	defer src.Close()
+
+	// Download to a temp file first then rename it to avoid partial download.
+	f, err := ioutil.TempFile(dir, file+".tmp")
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		if err := removeFileIfExists(f.Name()); err != nil {
+			log.Warnf("Error removing file: %s", err)
+		}
+	}()
+
+	if _, err := io.Copy(f, src); err != nil {
+		return err
+	}
+
+	if err := f.Close(); err != nil {
+		return err
+	}
+
+	// Dest is the final path of the disk image
+	dest := filepath.Join(dir, file)
+
+	// Windows can't rename in place, so remove the old file before
+	// renaming the temporary downloaded file.
+	if err := removeFileIfExists(dest); err != nil {
+		return err
+	}
+
+	return os.Rename(f.Name(), dest)
+}
+
+
+type B2dUtils struct {
+	releaseGetter
+	disk
+	storePath    string
+	imgCachePath string
+}
+
+func NewB2dUtils(storePath string) *B2dUtils {
+	imgCachePath := filepath.Join(storePath, "cache")
+
+	return &B2dUtils{
```

Comment:
messy to have to handle it this way.

would it be possible to strip this in future even more?